prog/vnet/cert_nexus.rb

```diff
@@ -0,0 +1,121 @@
+  subject_is :cert
+  semaphore :destroy
+
+  def self.assemble(hostname, dns_zone_id)
+    unless DnsZone[dns_zone_id]
+      fail "Given DNS zone doesn't exist with the id #{dns_zone_id}"
+    end
+
+    DB.transaction do
+      cert = Cert.create_with_id(hostname: hostname, dns_zone_id: dns_zone_id)
+
+      Strand.create(prog: "Vnet::CertNexus", label: "start") { _1.id = cert.id }
+    end
+  end
+
+  def before_run
+    when_destroy_set? do
+      hop_destroy unless %w[destroy].include?(strand.label)
+    end
+  end
+
+  label def start
+    register_deadline(:wait, 10 * 60)
+
+    private_key = OpenSSL::PKey::RSA.new(4096)
+    client = Acme::Client.new(private_key: private_key, directory: Config.acme_directory)
+    account = client.new_account(contact: "mailto:#{Config.acme_email}", terms_of_service_agreed: true, external_account_binding: {kid: Config.acme_eab_kid, hmac_key: Config.acme_eab_hmac_key})
+    order = client.new_order(identifiers: [cert.hostname])
+    authorization = order.authorizations.first
+    cert.update(kid: account.kid, private_key: private_key.to_s, order_url: order.url)
+    dns_challenge = authorization.dns
+    dns_zone.insert_record(record_name: "#{dns_challenge.record_name}.#{cert.hostname}", type: dns_challenge.record_type, ttl: 600, data: dns_challenge.record_content)
+
+    hop_wait_dns_update
+  end
+
+  label def wait_dns_update
+    dns_record = DnsRecord[dns_zone_id: dns_zone.id, name: "#{dns_challenge.record_name}.#{cert.hostname}.", tombstoned: false]
+    if DB[:seen_dns_records_by_dns_servers].where(dns_record_id: dns_record.id).all.empty?
+      nap 10
+    end
+
+    dns_challenge.request_validation
+
+    hop_wait_dns_validation
+  end
+
+  label def wait_dns_validation
+    case dns_challenge.status
+    when "pending"
+      nap 10
+    when "valid"
+      csr_key = OpenSSL::PKey::RSA.new(4096)
+      csr = Acme::Client::CertificateRequest.new(private_key: csr_key, common_name: cert.hostname)
+      acme_order.finalize(csr: csr)
+      cert.update(csr_key: csr_key.to_s)
+
+      hop_wait_cert_finalization
+    else
+      fail "DNS validation failed"
+    end
+  end
+
+  label def wait_cert_finalization
+    case acme_order.status
+    when "processing"
+      nap 10
+    when "valid"
+      cert.update(cert: acme_order.certificate, created_at: Time.now)
+
+      dns_zone.delete_record(record_name: "#{dns_challenge.record_name}.#{cert.hostname}")
+      hop_wait
+    else
+      fail "Certificate finalization failed"
+    end
+  end
+
+  label def wait
+    if cert.created_at < Time.now - 60 * 60 * 24 * 30 * 3 # 3 months
+      hop_destroy
+    end
+
+    nap 60 * 60 * 24 * 30 # 1 month
+  end
+
+  label def destroy
+    begin
+      acme_client.revoke(certificate: cert.cert) if cert.cert
+    rescue Sequel::Error
+      fail "Failed to revoke certificate"
+    end
+
+    dns_zone.delete_record(record_name: "#{dns_challenge.record_name}.#{cert.hostname}") if dns_challenge
```

Comment:
I'm not sure when `dns_challenge` can be null, but it seems we need to verify the existence of the DNS record and delete it regardless.

Reply:
I think the more pressing concern is if record_name is ever nilable, or hostname ever becomes nilable, and how we supervise whether that happens and what the downside risk is if t does happen (whereas, some other programming styles, like using `+`, would crash)

Reply:
The `dns_challenge` can be `nil` if the cert starts to be destroyed before the `dns_challenge` is requested. But if the `dns_challenge` is not `nil`, we must have the `record_name`, I'll make the necessary changes.